c10/core/DispatchKey.h

```diff
@@ -416,6 +418,7 @@ enum class DispatchKey : uint16_t {
   _QuantizedHIP,
   _QuantizedXLA,
   _QuantizedMLC,
+  _QuantizedIPU,
```

Comment:
There have been some recent changes to the DispatchKey representation, and there are now a few questions to think through when adding a new dispatch key.

Does IPU currently have existing support (existing kernels) for quantization? Sparsity? Does it need to override any autograd formulas from core?

If the answer to all of those is no, then you probably don't need to assign a `BackendComponent` enum slot for `IPU`. That's also the case for e.g. `ORT`, `Metal` and `Vulkan` dispatch keys, which don't currently override any of those functionalities (they just have backend kernels)

Reply:
I probably misunderstood the code: I thought the `_` prefixed enums were placeholders rather than implementations?

To answer your questions: yes we override some formulas for Autograd.
For Quantization / Sparsity it's too early to tell (Which is why I used the `_` version)

Reply:
C.f https://github.com/pytorch/pytorch/pull/74763/files/23d0af1a32c8c94f59fd0fa0280638ceb38dec42#diff-8b18203fde3227caea34b6078124836a22aca55e94d3d2b04fe781145adbddbbL409-L410

Reply:
>  yes we override some formulas for Autograd.

Just took a quick look at the poptorch repo - is the autograd registration mostly do deal with convolution? ([here](https://github.com/graphcore/poptorch/blob/ab184c575b52fd24f4621ef464f8db4e87a3cf3e/poptorch/source/dispatch_tracer/RegisterAtenOverloads.cpp#L378))

There has actually been a bunch of work (from @jbschlosser) to clean up the convolution ops in our aten namespace, and today we have two main convolution entry-points:

`at::convolution` is a non-composite op that you can directly write a backend kernel for with the backend `IPU` key ([here](https://github.com/pytorch/pytorch/blob/2ecc59086a5bc896c82f6f309d25569ec2933467/aten/src/ATen/native/native_functions.yaml#L1313))

And it's derivative formula directly calls the `at::convolution_backward` op ([here](https://github.com/pytorch/pytorch/blob/2ecc59086a5bc896c82f6f309d25569ec2933467/tools/autograd/derivatives.yaml#L2078)), which you an also directly write a kernel for with the `IPU` key.

I'm mostly just wondering if convolution was the only case where you had to override autograd (more because of the way that convolution was structured in core than anything specific to IPU's). cc @albanD.

Reply:
We haven't gone very far with the testing of the training part of the integration (Which is why for now it's only convolution, but you're right we had noticed convolution had changed in 1.11, so this will go away), but we think we'll potentially still need an Autograd key in 2 cases:
- For ops using the random numbers we need to make sure the prng reference is the same between forward and backward. (This is one example, but we might need to track more states to get correct and efficient behaviour).
- If we add custom operations which don't exist in pytorch then we need to register the autograd function against the autograd key? (We haven't tried that part yet).

Reply:
Thanks for the info! I'm less familiar with the rng example, but for point 2: if you have custom operators, and you want to define backward formulas for them, then yep you'll need an `AutogradIPU` key.